google/resource_dns_record_set.go

```diff
@@ -55,6 +57,147 @@ func resourceDnsRecordSet() *schema.Resource {
 	}
 }
 
+// This catches most invalid DNS names, though it might not catch some invlaid non-ASCII names
+// it does not implement the TOASCII functionality defined in https://tools.ietf.org/html/rfc3454
+// See:
+// https://tools.ietf.org/html/rfc1034 DOMAIN NAMES - CONCEPTS AND FACILITIES
+// https://tools.ietf.org/html/rfc3490 Internationalizing Domain Names in Applications (IDNA)
+// https://tools.ietf.org/html/rfc3454 Preparation of Internationalized Strings ('stringprep')
+func ValidateDNSRRsetName(i interface{}, k string) (s []string, errors []error) {
+	name := i.(string)
+	errors = append(errors, checkName(name)...)
+	return
+}
+
+func checkName(name string) (errors []error) {
+
+	if len(name) > 255 {
+		errors = append(errors, fmt.Errorf("DNS name length %d greater than 254",
+			len(name)))
+	}
+
+	// https://tools.ietf.org/html/rfc3490#section-3.1 Requirements
+	// Whenever dots are used as label separators, the following
+	// characters MUST be recognized as dots: U+002E (full stop),
+	// U+3002 (ideographic full stop), U+FF0E (fullwidth full
+	// stop), U+FF61 (halfwidth ideographic full stop).
+	re, err := regexp.Compile("[\u002e\u3002\uFF0E\uFF61]$")
+	if err != nil {
+		errors = append(errors, fmt.Errorf("Internal error compiling regexp: %s",
+			err))
+		re = nil
+	}
+	if re != nil && !re.MatchString(name) {
+		errors = append(errors,
+			fmt.Errorf("DNS name must be fully qualified with a trailing dot"))
+	}
+
+	dot_re, err := regexp.Compile("[\u002e\u3002\uFF0E\uFF61]")
+	if err != nil {
+		errors = append(errors, fmt.Errorf("Internal error compiling regexp: %s",
+			err))
+		dot_re = nil
+	}
+
+	verboten_re, err := regexp.Compile("[\x00-\x2C\x2E-\x2F\x3A-\x40\x5B-\x60\x7B-\x7F]+")
+	if err != nil {
+		errors = append(errors, fmt.Errorf("Internal error compiling regexp: %s",
+			err))
+		verboten_re = nil
+	}
+
+	all_digit_re, err := regexp.Compile("^[[:digit:]]+$")
+	if err != nil {
+		errors = append(errors, fmt.Errorf("Internal error compiling regexp: %s",
+			err))
+		all_digit_re = nil
+	}
+
+	space_re, err := regexp.Compile("[[:space:]]+")
+	if err != nil {
+		errors = append(errors, fmt.Errorf("Internal error compiling regexp: %s",
+			err))
+		space_re = nil
+	}
+
+	underscore_re, err := regexp.Compile("_+")
+	if err != nil {
+		errors = append(errors, fmt.Errorf("Internal error compiling regexp: %s",
+			err))
+		underscore_re = nil
+	}
+
+	var labels []string
+	if dot_re != nil {
+		labels := dot_re.Split(name, -1)
+		if len(labels) < 2 {
+			errors = append(errors,
+				fmt.Errorf("DNS name must not be a top level domain."))
+			labels = nil
+		}
+	}
+
+	for index, label := range labels {
+		// The final label is empty
+		if index+1 == len(labels) && len(label) < 1 {
+			errors = append(errors, nil)
```

Comment:
I'm a bit confused about why nil is being added to the errors list

Reply:
Err. Thinko. It made sense before this became a catch-all-errors compile time check. Fixed.